### PR TITLE
[Data] [Docs] Docs for working with PyTorch

### LINKED ISSUE #31418: [AIR] [Data] Guide to move from DataLoader to Ray Data
### Description

We have a guide to move existing Pytorch training code to Ray AIR TorchTrainer (https://docs.ray.io/en/latest/ray-air/examples/convert_existing_pytorch_code_to_ray_air.html).

However one aspect that is missing is specifically the data ingest part- how to move from PyTorch DataLoader to Ray Datasets (and what value does Ray Datasets provide?). Specifically for CV workloads.

Generally, pytorch data ingest looks like the following:
1. A list of S3 paths to images/labels. Or a parquet file with the following columns: S3 paths to images, label, additional metadata
2. Inside the training loop, shuffle the datasets for each epoch. Optionally sample the dataset based on the metadata.
3. Inside the training loop, for each batch, do the following: Load the images from S3. Preprocess the images. Apply random per-epoch augmentations.

We should show how to do the above setup with AIR ingest APIs and map from torch dataloader concepts to Ray AIR/Data concepts:
1. Read the parquet file
2. Use stream api and global shuffling in AIR DatasetConfig.
3. Specify a global preprocessor
4. Specify per-epoch preprocessor.

Specifically, the value adds that Ray Data provides:
1. Overlapping CPU preprocessing with GPU training.
2. Scale out CPU preprocessing independently from GPU training (using more CPU only nodes for preprocessing).

### Link

_No response_

### PR DESCRIPTION
**Pull request by amogkam**
Adds a guide for how PyTorch users can use Ray Data. This includes a migration guide from Torch Dataset/DataLoader to Ray Data. Closes https://github.com/ray-project/ray/issues/31418

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/data/api/from_other_data_libs.rst [hunk 1]
```diff
@@ -86,3 +86,9 @@ For PyArrow Users
      - :meth:`ds.groupby() <ray.data.Dataset.groupby>`
    * - pa.Table.sort_by()
      - :meth:`ds.sort() <ray.data.Dataset.sort>`
+
+
+For PyTorch Datasets & DataLoader Users
```
Comment (amogkam) [inline doc/source/data/api/from_other_data_libs.rst:91]:
```suggestion
For PyTorch Dataset & DataLoader Users
```

File: doc/source/data/working-with-pytorch.rst [hunk 2]
```diff
@@ -0,0 +1,505 @@
...
+
+    {'image': tensor([[[[...]]]], dtype=torch.uint8)}
+    ...
+    {'image': tensor([[[[...]]]], dtype=torch.uint8)}
+
+Integration with Ray Train
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Ray Data also integrates with :ref:`Ray Train <train-docs>` for easy data ingest for data parallel training, with support for PyTorch, PyTorch Lightning, or Huggingface training.
+
+See the :ref:`Ray Train user guide <train-datasets>` for more details.
```
Comment (bveeramani) [inline doc/source/data/working-with-pytorch.rst:45]: In the other guides and "Getting started", we put "For more details..." after the code snippet. Should we do the same here? Tradeoff I see is: 

- Advantage: Looks cleaner
- Disadvantage: Cross reference less visible

Don't have a strong opinion but consistency is nice

File: doc/source/data/working-with-pytorch.rst [hunk 3]
```diff
@@ -0,0 +1,505 @@
...
+    .. tab-item:: map_batches
+
+        .. testcode::
+            
+            from typing import Dict
+            import numpy as np
+            import torch
+            import ray
+            
+
```
Comment (bveeramani) [inline doc/source/data/working-with-pytorch.rst:141]: Remove newline for consistency with other tab
```suggestion

```

File: doc/source/data/working-with-pytorch.rst [hunk 4]
```diff
@@ -0,0 +1,505 @@
...
+    :options: +MOCK
+
+    {'output': array([0.5590901], dtype=float32)}
+
+.. _saving_pytorch:
+
+Saving Datasets containing PyTorch Tensors
+------------------------------------------
+
+Datasets containing torch tensors can be saved to files, like parquet or numpy.
```
[the remark this answers is not in the document]
Comment (amogkam) [reply doc/source/data/working-with-pytorch.rst:302]: whats wrong with other formats?

File: doc/source/data/working-with-pytorch.rst [hunk 5]
```diff
@@ -0,0 +1,505 @@
+.. _working_with_pytorch:
+
+Working with PyTorch
+====================
+
+Ray Data integrates with the PyTorch ecosystem.
+
+This guide describes how to:
+
+* :ref:`Iterate over your dataset as torch tensors for model training <iterating_pytorch>`
```
[the remark this answers is not in the document]
Comment (amogkam) [reply doc/source/data/working-with-pytorch.rst:10]: updated to use PyTorch when referencing the framework, and torch when referencing the package

File: doc/source/data/working-with-pytorch.rst (hunk 2 above)
Comment (amogkam) [reply doc/source/data/working-with-pytorch.rst]: Updated to adding references below the code snippets.

File: doc/source/data/working-with-pytorch.rst [hunk 6]
```diff
@@ -0,0 +1,505 @@
...
+    import torchvision
+    import ray
+
+    mnist = torchvision.datasets.MNIST(root="/tmp/", download=True)
+    ds = ray.data.from_torch(mnist)
+
+.. testoutput::
+    :hide:
+
+    ...
```
[the remark this answers is not in the document]
Comment (amogkam) [reply doc/source/data/working-with-pytorch.rst:363]: should we get rid of it then? I think it's still useful for quickly working with built-in datasets. Don't think anyone will want to port over all the logic for the built-in datasets. I will clarify what the schema looks like.

[the remark this answers is not in the document]
Comment (amogkam) [reply doc/source/data/working-with-pytorch.rst]: updated to show the schema

File: doc/source/data/working-with-pytorch.rst (hunk 4 above)
[the remark this answers is not in the document]
Comment (bveeramani) [reply doc/source/data/working-with-pytorch.rst:299]: They error, right?

```
>>> ray.data.range_tensor(1).write_csv("/tmp/data")
Traceback (most recent call last):       
...                                                                                                  ...
pyarrow.lib.ArrowInvalid: Unsupported Type:extension<arrow.py_extension_type<ArrowTensorType>>
```

[the remark this answers is not in the document]
Comment (bveeramani) [reply doc/source/data/working-with-pytorch.rst:299]: My understanding is that Parquet and NumPy are the only formats that can represent `ArrowTensorType`.

File: doc/source/data/working-with-pytorch.rst (hunk 6 above)
[the remark this answers is not in the document]
Comment (bveeramani) [reply doc/source/data/working-with-pytorch.rst]: I think we should get rid of it, but I guess that's orthogonal to this PR

**Review by bveeramani — APPROVED** [review]
LGTM except for the saving datasets section. If we can write to formats other than Parquet and NumPy, it's good as is. Otherwise, let's update it

File: doc/source/data/working-with-pytorch.rst (hunk 4 above)
[the remark this answers is not in the document]
Comment (bveeramani) [reply doc/source/data/working-with-pytorch.rst:299]: Discussed offline -- JSON also works

**Review by matthewdeng — APPROVED** [review]
stamp

**Pull request merged**